Use reuse tool 4.0

jonathanpallant opened the pull request:
This change upgrades us to reuse-tool 4.0.3, which has a new TOML format configuration instead of the old `.reuse/dep5` Debian-style file.

* Updated requirements file to install reuse-4.0.3
* Ran `reuse convert-dep5` to switch to new file format
* Switched over to `override` so the `REUSE.toml` file takes precedence over whatever random Copyright strings `reuse` finds in the source tree.

Should fix https://github.com/rust-lang/rust/issues/127361

jonathanpallant committed:
Upgrade to reuse 4.0.3

) Updated requirements file
) Ran `reuse convert-dep5` to switch to new file format

jonathanpallant committed:
Reformatted REUSE.toml

jonathanpallant committed:
Set all reuse annotations to override.

Kobzol commented on src/ci/docker/host-x86_64/mingw-check/reuse-requirements.txt:
```diff
@@ -1,9 +1,13 @@
 #
-# This file is autogenerated by pip-compile with Python 3.10
+# This file is autogenerated by pip-compile with Python 3.12
```
We use Python 3.10 in the Docker image of `mingw-check`. Could you please regenerate this lockfile with Python 3.10? Just to make sure that there are no surprises regarding Python package versions.

jonathanpallant replied:
Done. The hashes were unchanged.

Kobzol reviewed (commented):
This is great, thanks for working on it! I saw that you already discussed this with Pietro on the Ferrocene repo, and he is the author of the original issue, so he might want to bless this.

I cross-checked the TOML file with the original reuse file, everything looks fine. So apart from one nit (left a comment) LGTM.

r? @pietroalbini

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9994680199/job/27625055978) [(plain)](https://github.com/rust-lang/rust/commit/3b70d1be7f7394c05b6b3d72b913bdd3970aae35/checks/27625055978/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
reuse==4.0.3 \
    --hash=sha256:4f2c3e1213ec644e5febc50d8322d18982e4e1102af8a51d9493bfc2164a0eac \
    --hash=sha256:b33e26ec1d105cfcfc2e904d103faec0d758994278feb95a4f4290a864562243
    # via -r reuse-requirements.in
tomlkit==0.13.0 \
    --hash=sha256:08ad192699734149f5b97b45f1f18dad7eb1b6d16bc72ad0c2335772650d7b72 \
    # via reuse
#!/bin/bash
# Checks that no error code explanation is removed.

---
#12 [ 5/11] RUN npm install es-check@6.1.1 eslint@8.6.0 -g
#12 5.727 npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
#12 5.781 npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
#12 5.790 npm WARN deprecated rimraf@3.0.2: Rimraf versions prior to v4 are no longer supported
#12 5.814 npm WARN deprecated @humanwhocodes/config-array@0.9.5: Use @eslint/config-array instead
#12 5.843 npm WARN deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
#12 5.854 npm WARN deprecated @humanwhocodes/object-schema@1.2.1: Use @eslint/object-schema instead
#12 6.860 
#12 6.860 added 205 packages in 7s
#12 6.860 
#12 6.860 20 packages are looking for funding
---
#16 2.541   Getting requirements to build wheel: started
#16 2.605   Getting requirements to build wheel: finished with status 'done'
#16 2.607   Preparing metadata (pyproject.toml): started
#16 2.791   Preparing metadata (pyproject.toml): finished with status 'done'
#16 2.824 Collecting tomlkit==0.13.0
#16 2.831   Downloading tomlkit-0.13.0-py3-none-any.whl (37 kB)
#16 2.841   Building wheel for reuse (pyproject.toml): started
#16 3.085   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.086   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 3.086   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 3.087   Stored in directory: /tmp/pip-ephem-wheel-cache-9uya0qv2/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.089 Successfully built reuse
#16 3.089 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.481 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 DONE 3.6s

#17 [10/11] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#17 DONE 0.0s
---
* Bad licenses: 0
* Deprecated licenses: 0
* Licenses without file extension: 0
* Missing licenses: 0
* Unused licenses: CC0-1.0, CC-BY-3.0
* Used licenses: CC-BY-SA-4.0, OFL-1.1, LLVM-exception, MIT, Unicode-DFS-2016, BSD-2-Clause, NCSA, Apache-2.0
* Files with copyright information: 203380 / 203381
* Files with license information: 203380 / 203381

Unfortunately, your project is not compliant with version 3.2 of the REUSE Specification :-(
Unfortunately, your project is not compliant with version 3.2 of the REUSE Specification :-(


# RECOMMENDATIONS

* Fix unused licenses: At least one of the license text files in 'LICENSES' is
  not referenced by any file, e.g. by an 'SPDX-License-Identifier' tag. Please
  make sure that you either tag the accordingly licensed files properly, or
  delete the unused license text if you are sure that no file or code snippet is
  licensed as such.
* Fix missing copyright/licensing information: For one or more files, the tool
  cannot find copyright and/or licensing information. You typically do this by
  adding 'SPDX-FileCopyrightText' and 'SPDX-License-Identifier' tags to each
  file. The tutorial explains additional ways to do this:
  <https://reuse.software/tutorial/>
  local time: Thu Jul 18 16:14:38 UTC 2024
  network time: Thu, 18 Jul 2024 16:14:38 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

jonathanpallant committed:
Regenerated hashes using python3.10

) brew install python@3.10
) python3.10 -m venv /tmp/myenv
) source /tmp/myenv/bin/activate
) pip install pip-tools
) /tmp/myenv/bin/pip-compile --allow-unsafe --generate-hashes reuse-requirements.in

jonathanpallant committed:
Include REUSE.toml in REUSE.toml.

Also remove to licenses we no longer need, and given REUSE.toml reads some basic idea of what it's about.

pietroalbini commented:
Can you also revert https://github.com/rust-lang/rust/commit/c6eb03ba90f3d016a7fc75188182c08ce8281c40? With `precendence = "override"` set for LLVM, there should be no need for that manual tweaking anymore :D

jonathanpallant committed:
Removed CondensedDirectory support from license tools.

Now that we have reuse-tool 4.0, we no longer need to massage the JSON license data to collapse LLVM into a single copyright notice and license - reuse-tool can do it for us using an annotation in REUSE.toml.

This effectively reverts c6eb03b.

jonathanpallant commented:
That change has now been removed.

pietroalbini commented:
@bors r+

bors commented:
:pushpin: Commit 2e971bf5c69b3997fddab4c4c359ac3773d06247 has been approved by `pietroalbini`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=pietroalbini 2e971bf5c69b3997fddab4c4c359ac3773d06247 -->
<!-- homu: {"type":"Approved","sha":"2e971bf5c69b3997fddab4c4c359ac3773d06247","approver":"pietroalbini","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
